experimental/metal/direct_command_buffer.m

```diff
@@ -0,0 +1,1187 @@
+    iree_host_size_t offset, const void* values, iree_host_size_t values_length) {
+  iree_hal_metal_command_buffer_t* command_buffer =
+      iree_hal_metal_command_buffer_cast(base_command_buffer);
+
+  // "Binding a pipeline with a layout that is not compatible with the push constant layout does not
+  // disturb the push constant values." So we don't need to check whether the pipeline layout
+  // compatibility and invalidate existing values.
+  // See https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkCmdPushConstants.html
+
+  if (IREE_UNLIKELY(offset + values_length >= sizeof(command_buffer->state.push_constants))) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "push constant range [%zu, %zu) out of range", offset,
+                            offset + values_length);
+  }
+
+  memcpy((uint8_t*)&command_buffer->state.push_constants + offset, values, values_length);
+  command_buffer->state.current_pipeline_layout = pipeline_layout;
+
+  return iree_ok_status();
+}
+
+static int compare_descriptor(const void* a, const void* b) {
+  const iree_hal_metal_descriptor_t* buffer_a = (const iree_hal_metal_descriptor_t*)a;
+  const iree_hal_metal_descriptor_t* buffer_b = (const iree_hal_metal_descriptor_t*)b;
+  if (buffer_a->set != buffer_b->set) return buffer_a->set - buffer_b->set;
+  return buffer_a->binding - buffer_b->binding;
+}
+
+// Returns true if the given |descriptors| array contains descriptors in ascending binding slot
+// order and there is no duplicated binding slots.
+static bool iree_hal_metal_is_sorted_unique_descriptors(iree_hal_metal_descriptor_t* descriptors,
+                                                        int descriptor_count) {
+  for (int i = 1; i < descriptor_count; ++i) {
+    if (compare_descriptor(&descriptors[i - 1], &descriptors[i]) >= 0) return false;
+  }
+  return true;
+}
+
+static inline MTLResourceUsage iree_hal_metal_get_metal_resource_usage(
+    iree_hal_descriptor_set_layout_binding_t* binding) {
+  MTLResourceUsage usage = MTLResourceUsageRead;
+  if (binding->flags != IREE_HAL_DESCRIPTOR_FLAG_READ_ONLY) usage |= MTLResourceUsageWrite;
+  return usage;
+}
+
+static iree_status_t iree_hal_metal_command_buffer_push_descriptor_set(
+    iree_hal_command_buffer_t* base_command_buffer, iree_hal_pipeline_layout_t* pipeline_layout,
+    uint32_t set, iree_host_size_t binding_count,
+    const iree_hal_descriptor_set_binding_t* bindings) {
+  iree_hal_metal_command_buffer_t* command_buffer =
+      iree_hal_metal_command_buffer_cast(base_command_buffer);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  IREE_ASSERT(set != IREE_HAL_METAL_PUSH_CONSTANT_BUFFER_INDEX);
+  iree_hal_metal_descriptor_t* descriptors = command_buffer->state.current_descriptors;
+  IREE_ASSERT(iree_hal_metal_is_sorted_unique_descriptors(
+      descriptors, command_buffer->state.current_total_binding_count));
```

Comment:
Asserts intended? Are these really never possible, or should this return an error instead?

Reply:
Yes. Though this does not matter anymore--I changed how those descriptors are handled so no need to sort anymore.